processing/src/main/java/io/druid/segment/data/GenericIndexedWriter.java

```diff
@@ -158,10 +323,111 @@ public InputStream getInput() throws IOException
+      int filesRequired = (int) Math.ceil((float) numWritten / bagSize);
+
+      try (InputStream is = new FileInputStream(ioPeon.getFile(makeFilename("values")))) {
+        int counter = -1;
+        byte[] buffer = new byte[1 << 16];
+
+        for (int i = 0; i < filesRequired; i++) {
+          if (i != filesRequired - 1) {
+            headerFile.seek((bagSize + counter) * Longs.BYTES); // 8 for long bytes.
+            counter = counter + bagSize;
+          } else {
+            headerFile.seek((numWritten - 1) * Longs.BYTES); // for remaining items.
+          }
+
+          long valuePosition = headerFile.readLong();
+          long numBytesToPutInFile = valuePosition - previousValuePosition;
+
+          try (SmooshedWriter smooshChannel = smoosher
+              .addWithSmooshedWriter(generateValueFileName(filenameBase, i), numBytesToPutInFile)) {
+            writeBytesIntoSmooshedChannel(numBytesToPutInFile, buffer, smooshChannel, is);
+            previousValuePosition = valuePosition;
+          }
+        }
+      }
+      writeHeader(smoosher, headerFile, bagSizePower);
+    }
+  }
+
+  private void writeHeader(FileSmoosher smoosher, RandomAccessFile headerFile, int bagSizePower) throws IOException
+  {
+    try (CountingOutputStream finalHeaderOut = new CountingOutputStream(
+        ioPeon.makeOutputStream(makeFilename("header_final")))) {
+      int bagSize = 1 << bagSizePower;
+      long currentNumBytes = 0;
+      long relativeRefBytes = 0;
+      long relativeNumBytes;
+
+      for (int pos = 0; pos < numWritten; pos++) {
+        headerFile.seek(pos * Longs.BYTES);
+        currentNumBytes = headerFile.readLong();
+        relativeNumBytes = currentNumBytes - relativeRefBytes;
+        finalHeaderOut.write(Ints.toByteArray((int) (relativeNumBytes)));
+
+        if (--bagSize == 0) {
+          relativeRefBytes = currentNumBytes;
+          bagSize = 1 << bagSizePower;
+        }
+      }
+
+      long numBytesToPutInFile = finalHeaderOut.getCount();
+      finalHeaderOut.close();
+      byte[] buffer = new byte[1 << 16];
+      try (InputStream is = new FileInputStream(ioPeon.getFile(makeFilename("header_final")))) {
+        try (SmooshedWriter smooshChannel = smoosher
+            .addWithSmooshedWriter(generateHeaderFileName(filenameBase), numBytesToPutInFile)) {
+          writeBytesIntoSmooshedChannel(numBytesToPutInFile, buffer, smooshChannel, is);
+        }
+      }
+
+    }
+  }
+
+  private void initializeHeaderOutLong() throws IOException
   {
-    try (final ReadableByteChannel from = Channels.newChannel(combineStreams().getInput())) {
-      ByteStreams.copy(from, channel);
+    headerOut.close();
+    headerOutLong = new CountingOutputStream(ioPeon.makeOutputStream(makeFilename("headerLong")));
+    File f = ioPeon.getFile(makeFilename("header"));
+
+    try (RandomAccessFile headerFile = new RandomAccessFile(f, "r")) {
+
+      for (int i = 0; i < numWritten; i++) {
```

Comment:
Seems that it could be convert into bulk transfer from `headerFile.getChannel()` to `headerOutLong`, without a lot of RAF seek() calls